### ISSUE TITLE
Wrong assumption regarding the structure of the DjVu text layer

### ISSUE THREAD

**Issue by baskerville** [koreader/koreader#2943]
*KOReader* assumes that the text layer of the DjVu files has the form `page -> lines -> words`. But it could as well be `page -> columns -> regions -> paragraphs -> lines -> words` or anything in between.

Here's an excerpt from `man djvused`:
> The  hidden  text  for each page is simply represented by a single structural element of type page.  Various level of structural information are acceptable. For instance, the page level component might only specify a page level string, or might only provide a list of lines, or might provide a full hierarchy down to the individual characters.

**Comment by Frenzie** [comment]
Pinging @xelxebar. Any interest? :-)

**Comment by xelxebar** [comment]
Oh boy. Interest, yes, but it will take me a good bit of time to get up to speed.

My PR was actually just hacked together on a broken dev environment, so I nuked it. If you don't mind me taking my time and badgering you with questions sometimes, then I'll gladly take this on!

First, I need to read those docs on properly setting up a build environment (virdevenv, I guess?).

**Comment by Frenzie** [comment]
Not a problem! 

What OS do you use?

**Comment by xelxebar** [comment]
Thanks! Here's my machine info:

```
$ lsb_release --all
LSB Version:	1.0
Distributor ID:	VoidLinux
Description:	Void Linux
Release:	rolling
Codename:	void

$ uname -srvm
Linux 4.18.19_1 #1 SMP PREEMPT Wed Nov 14 23:03:54 UTC 2018 x86_64
```

**Comment by Frenzie** [comment]
In that case you could just set up the build environment as per https://github.com/koreader/koreader#building-prerequisites or indeed use something like Vagrant or Docker. For Docker you'd probably want to play around with the `--mount` flag or make your own container with the Dockerfile that does it for you. I haven't really tried it that way.

**Comment by xelxebar** [comment]
Yeah, originally followed that but didn't use koxtoolchain, so my kindle target ended up crashing on the native hardware.

Anyway, I hate to spam up the issue comments with my unrelated issues, is there an IRC room or something we can migrate to?

**Comment by Frenzie** [comment]
There's [Gitter](https://gitter.im/koreader/koreader?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) or we could chat on Discord or IRC if you prefer.

**Comment by xelxebar** [comment]
@baskerville, I am finally getting around to looking at this, and want to make sure I understand the issue.

I'm guessing that you're referring to `djvu.c/getPageText` in base. This builds it's table from the result of `ddjvu_document_get_pagetext` which seems to correspond to a `TXTa/TXTz` section which, according to section `8.3.5` of the [DjVu v3 Specification][0] describes it as having the more freely-organized structure you mention.

So, just as you point out, `getPageText` looks like it should be returning some table capable of representing this more flexible layout.

Am I missing anything?

[0]:http://www.djvu.org/docs/DjVu3Spec.djvu

**Comment by baskerville** [comment]
@xelxebar You have a clear understanding of the problem.

**Comment by xelxebar** [comment]
@baskerville 
Did you notice this issue on an actual DjVu file? I've spent some time looking for files that make use of a more flexible hierarchy, but almost everyone seems to be using `page -> line -> word`.

I'm working on a script that will generate some more "exotic" DjVu for testing, but it would be helpful if you could share any examples you have.

Cheers!

**Comment by baskerville** [comment]
[hocr2djvused](http://jwilk.net/software/ocrodjvu) often generates *columns* and *regions*.

**Comment by xelxebar** [comment]
@baskerville 
Would you mind sharing any of these? The test djvu I'm creating have artificially-munged annotation hierarchies, but I'd really like some real-world examples to make sure text selection really is acting sanely.

**Comment by baskerville** [comment]
@xelebar I can only share it privately, so you'll need to send me an e-mail.

**Comment by xelxebar** [comment]
Hrm. I'm hoping to add it to `test/`, actually. Thanks though!

**Comment by vejkse** [comment]
This is a bit late, but here is a zipped (djvu isn’t an allowed file type) real-world example file, OCRed with ocrodjvu, which creates a `page -> column -> para -> line -> word` hierarchy.  When trying to select a word in koreader, we get a whole “column”.

[page-colum-para-line-word.zip](https://github.com/koreader/koreader/files/11202526/page-colum-para-line-word.zip)
